Update walkthrough with kubectl command examples

karlkfi opened the pull request:
The walkthrough was rather unactionable. So I added kubectl examples.
Hopefully it can actually be "used" now instead of just "read".

karlkfi commented:
I'm hoping to use this for a hack-day at gophercon, next week. So it'd be nice to get some eyes on it!

The idea is to use it after setting up a cluster with one of the getting started guides, like the new one for mesos/docker: https://github.com/GoogleCloudPlatform/kubernetes/pull/10049

karlkfi commented:
TODO: add details about using service proxy to access the service endpoint

zmerlynn commented:
The Shippable failures are real:
```
--- FAIL: TestExampleObjectSchemas (0.51s)
examples_test.go:340: skipping : ../examples/elasticsearch/apiserver-secret.yaml/apiserver-secret
examples_test.go:340: skipping : ../examples/multi-pod.yaml/multi-pod
examples_test.go:335: ../examples/walkthrough/pod-nginx-with-label.yaml: pod-nginx-with-label does not have a test case defined
examples_test.go:335: ../examples/walkthrough/pod-nginx.yaml: pod-nginx does not have a test case defined
examples_test.go:335: ../examples/walkthrough/pod-redis.yaml: pod-redis does not have a test case defined
examples_test.go:363: Expected 6 examples, Got 4
FAIL
coverage: 0.0% of statements
FAIL	github.com/GoogleCloudPlatform/kubernetes/examples	0.655s
```

RichieEscarez commented on examples/walkthrough/README.md:
```diff
@@ -1,73 +1,149 @@
+If you downloaded a pre-compiled release, kubectl should be under `platforms/<os>/<arch>`.
+
+If you built from source, kubectl should be either under `_output/local/bin/<os>/<arch>` or `_output/dockerized/bin/<os>/<arch>`.
+
```
Insert link to kubectl ref docs:

**Tip:** View all the `kubectl` commands, including their options and descriptions in the [kudectl CLI reference](../../docs/kubectl.md).

karlkfi replied:
addressed below

RichieEscarez commented on examples/walkthrough/README.md:
```diff
@@ -1,73 +1,149 @@
+
+If you built from source, kubectl should be either under `_output/local/bin/<os>/<arch>` or `_output/dockerized/bin/<os>/<arch>`.
+
+### Install
```
Make smaller and revise: 

####Installing kubectl

karlkfi replied:
Seems like a style complaint about how the github header levels don't look different enough. I'm not sure skipping a header level really solves the problem...

karlkfi replied:
I think I'll just change all the ### to #### for consistency in 101 and 201.

RichieEscarez commented on examples/walkthrough/README.md:
```diff
@@ -1,73 +1,149 @@
+Do one of the following to install kubectl:
+- Add the kubectl binary to your PATH
+- Move kubectl into a dir already in PATH (like `/usr/local/bin`)
+- Use `./cluster/kubectl.sh` instead of kubectl. It will auto-detect the location of kubectl and proxy commands to it.
```
Can we revise lines 12-16, so that whats at line 16 is moved above the install options (16 its not really "install" but more of an alternative):

####Installing kubectl
You can use the version of kubectl thats included in the directory where you installed Kubernetes or install kubectl by adding it to your PATH environment variable.

 -  Run the kubectl commands by using `./cluster/kubectl.sh`, for example:  
    `$ ./cluster/kubectl.sh cluster-info`  
    (*Note: It will auto-detect the location of kubectl and proxy commands to it.*)
 - Install kubectl by adding it to PATH so that you can simply use 'kubectl', for example:  
    `$ kubectl cluster-info`

RichieEscarez commented on examples/walkthrough/README.md:
```diff
@@ -1,73 +1,149 @@
+- Move kubectl into a dir already in PATH (like `/usr/local/bin`)
+- Use `./cluster/kubectl.sh` instead of kubectl. It will auto-detect the location of kubectl and proxy commands to it.
+
+### Configure
```
Also change to:
####Configuring kubectl

RichieEscarez commented on examples/walkthrough/README.md:
````diff
@@ -1,73 +1,149 @@
+If your cluster was deployed by other means, you may want to configure the path to the Kubernetes apiserver in your shell environment:
+
+```sh
+export KUBERNETES_MASTER=http://<ip>:<port>/api
````
prefix with "$" -> $ export .... 
(this is the convention im seeing across all other examples)
Please make this change throughout the doc for all commands.

karlkfi replied:
gotta say, I'm really not a fan of this convention, especially for code blocks tagged as being shell code. The prompt is not part of bash or shell code. The only time I prefer prefixing with a $ is when showing full console input and output, including stdout/stderr, to distinguish between input and output.

RichieEscarez commented on examples/walkthrough/README.md:
```diff
@@ -1,73 +1,149 @@
+
+By default, kubectl configuration lives at `~/.kube/config`.
+
+If your cluster was deployed by other means, you may want to configure the path to the Kubernetes apiserver in your shell environment:
```
Revise/insert: "....by other means (for example a [getting started guide](../../docs/getting-started-guides/README.md), ...."

RichieEscarez commented on examples/walkthrough/README.md:
````diff
@@ -1,73 +1,149 @@
+Check that kubectl is properly configured by getting the cluster state:
+
+```sh
+kubectl cluster-info
````
another, prefix with "$" -> $ kubectl .... 

(will stop calling these out, but please catch/correct the rest)

bgrant0607 replied:
+1

mikedanese replied:
$ implies you are on the command line as the guide is describing. +1

RichieEscarez commented on examples/walkthrough/README.md:
```diff
@@ -1,73 +1,149 @@
+
+### Managing a Pod
+
+If you have the source checked out, the following command (executed from the root of the project) will create the pod defined above. Otherwise download [the pod definition](pod-nginx.yaml) first.
```
A pre-compiled release also includes the "examples" directory. We should revise to refer to the "Kubernetes root directory".

RichieEscarez replied:
Plus remove reference to download. Maybe instead say "You can also view a copy of the [nginx pod definition here](pod-nginx.yaml)."

karlkfi commented:
Gotta say, I'm really not a fan of the `$` in the code block convention, especially for code blocks tagged as being shell code. The prompt is not part of bash or shell code. 

The usage I would conceded would be when giving an example of full console input and output, including stdout/stderr, to distinguish between input and output.

I can add them if you insist, of course, but I'd be more tempted to just go remove them everywhere else. ;)

Should it also be used for inline shell/command blocks?

karlkfi commented:
```
$ go test ./examples
ok  	github.com/GoogleCloudPlatform/kubernetes/examples	0.053s
```

RichieEscarez commented on examples/walkthrough/README.md:
````diff
@@ -1,73 +1,149 @@
+That's great for a simple static web server, but what about persistent storage?
+
+The container file system only lives as long as the container does. So if your app's state needs to survive reboots or crashes you'll need to configure some persistent storage.  To do so, declare a ```volume``` in the pod definition, and mount it into one or more of the pod containers:
+
````
Delete: 
"To do so, declare a ```volume``` in the pod definition, and mount it into one or more of the pod containers:"

And then on a new line, add:
"To configure persistent storage, edit the pod definition by declaring the volume and then mounting that volume in one or more of the pod containers:"

RichieEscarez commented on examples/walkthrough/README.md:
````diff
@@ -1,73 +1,149 @@
+
+The container file system only lives as long as the container does. So if your app's state needs to survive reboots or crashes you'll need to configure some persistent storage.  To do so, declare a ```volume``` in the pod definition, and mount it into one or more of the pod containers:
+
+Added a volume:
````
Make ordered list (1)

RichieEscarez commented on examples/walkthrough/README.md:
````diff
@@ -1,73 +1,149 @@
+    emptyDir: {}
+```
+
+Reference the volume in our container section:
````
2. Mount the volume by adding the following to the pod definition:

RichieEscarez commented on examples/walkthrough/README.md:
````diff
@@ -1,73 +1,149 @@
+      mountPath: /data/redis
+```
+
+Full definition ([source](pod-redis.yaml)):
````
Revise to:
Result: The definition for a Redis persistent storage pod ([pod-redis.yaml](pod-redis.yaml)):

RichieEscarez commented on examples/walkthrough/README.md:
````diff
@@ -1,73 +1,149 @@
+  - name: redis-persistent-storage
+    emptyDir: {}
 ```
 
````
Insert before lines 140-144:
Tips:

karlkfi replied:
not sure what this comment means....

RichieEscarez replied:
Sorry, the lines have moved due to edits but since the two statements about "emptydir" and "hostdir" provide additional information (vs. being actionable steps), im suggesting to insert "Tips:" to introduce the remaining portion of this section:

Tips:
In Kubernetes, ```emptyDir``` Volumes live for the lifespan of the Pod, which.....

If you want to mount a directory that already exists in the file systee.....

See.....

RichieEscarez commented:
Hi @karlkfi, sorry I failed to refresh the page and didnt notice your responses coming in. 

Im fine without the $'s (you made valid points). 

Thanks for making all the changes so far. Ill look over the 201 page next.

RichieEscarez commented on examples/walkthrough/k8s201.md:
```diff
@@ -1,26 +1,69 @@
-When we had just left off in the [previous episode](README.md) we had learned about pods, multiple containers and volumes.
-We'll now cover some slightly more advanced topics in Kubernetes, related to application productionization, deployment and
+If you went through [Kubernetes 101](README.md), you learned about kubectl, pods, volumes, and multiple containers.
+For Kubernetes 201 we'll cover some slightly more advanced topics in Kubernetes, related to application productionization, deployment and
```
To clarify that the examples here build off of the 101 page, please insert/revise to:

For Kubernetes 201, we will pick up where 101 left off and cover....

RichieEscarez commented on examples/walkthrough/k8s201.md:
````diff
@@ -48,15 +91,35 @@ spec:
 ```
 
-### Services
+To create such a replication controller:
````
Depending on how you format the pod management section in 101, this (lines 96-110) section (and also the similar "services" section, lines 137-166) should be formatted similarly.

In the 101 page i suggested to highlight these task and clarify them as "common key tasks" because as they are now, they are not introduced.

RichieEscarez commented on examples/walkthrough/k8s201.md:
````diff
@@ -147,7 +241,8 @@ spec:
         - containerPort: 80
 ```
 
````
Can we add a pointer here to the "Container Probes" section in the "Life of a Pod" page? 

For more information about health checking, see [Container Probes](../../docs/pod-states.md#container-probes).

RichieEscarez commented:
@karlkfi, Im done with my review of the two docs. Let me know when to take a final look. Thanks!

karlkfi commented:
@RichieEscarez all comments addressed.

karlkfi commented:
Test failure doesn't look related...

```
At beginning of loop: foo-v1 replicas: 0, foo-v2 replicas: 1
At end of loop: foo-v1 replicas: 0, foo-v2 replicas: 1
At beginning of loop: foo-v1 replicas: 1, foo-v2 replicas: 1
At end of loop: foo-v1 replicas: 1, foo-v2 replicas: 1
At beginning of loop: foo-v1 replicas: 0, foo-v2 replicas: 2
At end of loop: foo-v1 replicas: 0, foo-v2 replicas: 2
At beginning of loop: foo-v1 replicas: 1, foo-v2 replicas: 1
At end of loop: foo-v1 replicas: 1, foo-v2 replicas: 1
At beginning of loop: foo-v1 replicas: 0, foo-v2 replicas: 2
At end of loop: foo-v1 replicas: 0, foo-v2 replicas: 2
At beginning of loop: foo-v1 replicas: 6, foo-v2 replicas: 1
At end of loop: foo-v1 replicas: 6, foo-v2 replicas: 1
At beginning of loop: foo-v1 replicas: 5, foo-v2 replicas: 2
At end of loop: foo-v1 replicas: 5, foo-v2 replicas: 2
At beginning of loop: foo-v1 replicas: 1, foo-v2 replicas: 2
At end of loop: foo-v1 replicas: 1, foo-v2 replicas: 2
At beginning of loop: foo-v1 replicas: 0, foo-v2 replicas: 3
At end of loop: foo-v1 replicas: 0, foo-v2 replicas: 3
--- FAIL: TestUpdateWithRetries (0.68s)
rolling_updater_test.go:719: unexpected error: timed out waiting for the condition
rolling_updater_test.go:724: Remaining updates [0xc2081a3680] gets []
FAIL
coverage: 57.1% of statements
FAIL	github.com/GoogleCloudPlatform/kubernetes/pkg/kubectl	1.343s
```

sttts commented on examples/walkthrough/README.md:
```diff
@@ -1,74 +1,167 @@
+
+#### Install Kubectl
+
+The kubectl binary can be executed from anywhere, but the rest of the walkthrough will assume that it's in your PATH.
```
Sentence does not really make sense. If it is in the path, it can be run from anywhere. The "but" is strange.

karlkfi replied:
I meant that you could execute the binary from anywhere, even if it's not on the path, provided that you know the path, but I see that the phrasing is confusing.

sttts commented on examples/walkthrough/README.md:
````diff
@@ -1,74 +1,167 @@
+Create a pod containing an nginx server:
+
+```sh
+kubectl create -f examples/walkthrough/pod-nginx.yaml
````
Called just pod.yaml now.

karlkfi replied:
I renamed it in this PR.

sttts commented on examples/walkthrough/README.md:
```diff
@@ -1,74 +1,167 @@
+
+#### Multiple Containers
 
 _Note:
```
This sentence is confusing. Does it mean the examples are work in progress and don't work? Why are they here then?

karlkfi replied:
I can't speak for the author, but if you want to make them working examples feel free to make a new example. I'm not going to modify them much as part of this PR, I just fixed some of the page formatting.

bgrant0607 replied:
It means that nobody has built and pushed the required container image for the example, so the example doesn't work exactly as written.

sttts commented on examples/walkthrough/k8s201.md:
````diff
@@ -1,26 +1,73 @@
+    name: nginx
+```
+
+Full labeled nginx pod definition ([source](pod-nginx-with-label.yaml)):
````
A full ... looks like this:

sttts commented on examples/walkthrough/k8s201.md:
```diff
@@ -1,26 +1,73 @@
+## Replication Controllers
 
-OK, now you have an awesome, multi-container, labelled pod and you want to use it to build an application, you might be tempted to just start building a whole bunch of individual pods, but if you do that, a whole host of operational concerns pop up.  For example: how will you scale the number of pods up or down and how will you ensure that all pods are homogenous?
+OK, now you know how to make awesome, multi-container, labelled pods and you want to use them to build an application, you might be tempted to just start building a whole bunch of individual pods, but if you do that, a whole host of operational concerns pop up.  For example: how will you scale the number of pods up or down and how will you ensure that all pods are homogenous?
```
"labeled" or "labelled"? Both appear in the file.

karlkfi replied:
labeled is American English, labelled is British. I'll just make sure it's consistent.

sttts commented on examples/walkthrough/k8s201.md:
```diff
@@ -48,15 +95,40 @@ spec:
-### Services
+#### Replication Controller Management
+
+If you have the source checked out or downloaded the pre-compiled release you should have an `examples` directory in the Kubernetes root directory. If not, download the [nginx replication controller example definition](replication-controller.yaml).
```
This sentence does not appear above for the labelled pod. There it's assumes the example directory exists.

karlkfi replied:
Replacing all these duplicate lines with a comment at the top of the file.

sttts commented on examples/walkthrough/k8s201.md:
```diff
@@ -48,15 +95,40 @@ spec:
+
+## Services
+
 Once you have a replicated set of pods, you need an abstraction that enables connectivity between the layers of your application.  For example, if you have a replication controller managing your backend jobs, you don't want to have to reconfigure your front-ends whenever you re-scale your backends.  Likewise, if the pods in your backends are scheduled (or rescheduled) onto different machines, you ...
```
A "Service" or a "service". Both are mixed in this file.

sttts replied:
Better: ... basically like a simple load-balancer in front of that service with a stic IP.

sttts commented on examples/walkthrough/k8s201.md:
```diff
@@ -71,9 +143,43 @@ spec:
 
+#### Service Management
+
+If you have the source checked out or downloaded the pre-compiled release you should have an `examples` directory in the Kubernetes root directory. If not, download the [nginx service example definition](service.yaml).
```
Again this sentence is not consistently used in this file.

sttts commented on examples/walkthrough/k8s201.md:
````diff
@@ -71,9 +143,43 @@ spec:
+kubectl get services
+```
+
+Provided your pod & service IPs are accessible (depending on network setup), you should be able to access the service's http endpoint in a browser or curl on port 80.
````
For most people the network will not be setup as expected. It would be worth to follow kube-dns' example, compare https://github.com/GoogleCloudPlatform/kubernetes/tree/master/cluster/addons/dns#1-create-a-simple-pod-to-use-as-a-test-environment.

karlkfi replied:
This is complicated... I'm not sure I really want to get into all that in the walkthrough... I'd prefer a different page to link to that talks about using busybox, but I'm not sure I want to write one as part of this PR.

bgrant0607 replied:
In general, these addresses aren't going to be accessible. Use a cloud load balancer or the apiserver proxy, as suggested above. Or, you could use load balancer type NodePort.

sttts commented on examples/walkthrough/k8s201.md:
```diff
@@ -147,8 +254,12 @@ spec:
 
-### What's next?
-For a complete application see the [guestbook example](https://github.com/GoogleCloudPlatform/kubernetes/tree/master/examples/guestbook).
+For more information about health checking, see [Container Probes](/docs/pod-states.md#container-probes).
```
I prefer this sentence over "... explained elsewhere" above.

karlkfi replied:
agreed

RichieEscarez commented:
@karlkfi thanks for making all the changes! I did respond to the question about one of my comments and also see more comments from @sttts. LGTM pending the recovery of the remaining open comments.

karlkfi commented:
bah... this sat so long there's been a half dozen other changes in master that conflict....

karlkfi commented:
Ok, all outstanding comments addressed, please review, @RichieEscarez

karlkfi commented:
Would love to get this in 1.0. Are docs changes still getting cherry-picked over? @davidopp

davidopp commented:
I know we're still accepting stuff under docs/ but not sure about examples/
I'll check and get back to you

davidopp commented:
The verdict is that we're still accepting changes to examples/

Looks like this is assigned to @bgrant0607 , I'll let him review it.

karlkfi commented:
rebased to fix mungedocs shippable failures

bgrant0607 commented on examples/walkthrough/README.md:
```diff
@@ -13,76 +14,175 @@ certainly want the docs that go with that version.</h1>
 
 <!-- END MUNGE: UNVERSIONED_WARNING -->
 # Kubernetes 101 - Walkthrough
+=======
```
Unresolved merge conflict

karlkfi replied:
huh, wonder how i missed that...

bgrant0607 commented on examples/walkthrough/README.md:
```diff
@@ -13,76 +14,175 @@ certainly want the docs that go with that version.</h1>
+  - name: nginx
+    image: nginx
+    ports:
+    - name: http
```
If you want it to be even simpler, the port name is not required

bgrant0607 commented on examples/walkthrough/README.md:
````diff
@@ -13,76 +14,175 @@ certainly want the docs that go with that version.</h1>
+      containerPort: 80
 ```
 
 A pod definition is a declaration of a _desired state_.  Desired state is a very important concept in the Kubernetes model.  Many things present a desired state to the system, and it is Kubernetes' responsibility to make sure that the current state matches the desired state.  For example, when you create a Pod, you declare that you want the containers in it to be running.  If the containers happe ...
````
rephrase last sentence: "The process continues until the Pod is deleted."

(other things can cause the pod to be deleted)

bgrant0607 commented on examples/walkthrough/README.md:
````diff
@@ -13,76 +14,175 @@ certainly want the docs that go with that version.</h1>
+kubectl get pods
+```
+
+Provided your pod IPs are accessible (depending on network setup), you should be able to access the pod's http endpoint in a browser or curl on port 80.
````
This usually isn't going to work. 

The proper way to do this is either with a service with a cloud load balancer, or using the apiserver proxy:
https://github.com/GoogleCloudPlatform/kubernetes/blob/master/docs/accessing-the-cluster.md#manually-constructing-apiserver-proxy-urls

karlkfi replied:
It works on some providers. It just depends on how your networking is set up and where you're running the commands. Tho we have the same issue with services. The only relatively easy cross-provider way to do it is to use a busybox pod, but I don't want to have to inject that guide everywhere. So I'll probably make an external doc about it.

bgrant0607 replied:
The apiserver proxy works on every provider.

bgrant0607 replied:
It looks like this (and the corresponding service text) is the last issue that needs to be resolved before we merge this PR.

bgrant0607 commented on examples/walkthrough/README.md:
```diff
@@ -13,76 +14,175 @@ certainly want the docs that go with that version.</h1>
+
+For this example, we'll be creating a Redis pod, with a named volume and volume mount that defines the path to mount the volume.
+
+1. Define a volume:
```
I think this is hard to understand without seeing both config fragments in a complete pod configuration.

bgrant0607 replied:
Ah, see it below. Nevermind.

bgrant0607 commented on examples/walkthrough/README.md:
```diff
@@ -94,7 +194,7 @@ However, often you want to have two different containers that work together.  An
 kind: Pod
 metadata:
-  name: www
+  name: git-monitor
```
The pod isn't git-monitor. That's a helper container, not the primary container.

karlkfi replied:
`www` is a terribly generic name. what are we serving? `/srv/www` doesn't exist anywhere...

bgrant0607 commented on examples/walkthrough/k8s201.md:
```diff
@@ -12,29 +12,79 @@ certainly want the docs that go with that version.</h1>
+metadata:
+  name: nginx
+  labels:
+    name: nginx
```
Please don't use labels with the key "name". People find that confusing. I suggest "app", "component", or somesuch.

See #7257 for more best practices for examples.

bgrant0607 commented:
Thanks for the work on this @karlkfi.

bgrant0607 commented:
FYI, this will need a rebase after #11239 goes in. We're moving the walkthrough to docs/user-guide/walkthrough.

karlkfi commented:
now that's just mean... moving it out from under me! :P

karlkfi commented:
All comments addressed. Maybe you should just merge mine before #11239 ;)

mikedanese commented:
Uh oh :). Please rebase to master and squash the fixup commits.

karlkfi commented:
nooooooooooooo... (ok fine)

Commit:
Add command execution (cmd-exec) example using busybox

Commit:
Walkthrough example cleanup

- Add kubectl command examples
- Add tables of contents
- Skip 3rd header tier to make sections more clear
- Reference cmd-exec example for curling pod & service IPs
- Make section layout, text patterns, changes & links more consistent
- Canonical yaml formatting

karlkfi commented:
squashed and rebased.

bgrant0607 commented on docs/user-guide/walkthrough/README.md:
```diff
@@ -12,77 +12,176 @@ certainly want the docs that go with that version.</h1>
+
+In order for the kubectl usage examples to work, make sure you have an examples directory locally, either from [a release](https://github.com/GoogleCloudPlatform/kubernetes/releases) or [the source](https://github.com/GoogleCloudPlatform/kubernetes).
+
+Table of Contents
```
Please convert to the standard TOC style. For example: https://github.com/GoogleCloudPlatform/kubernetes/blob/master/docs/user-guide/configuring-containers.md

karlkfi replied:
Are there any docs on how cmd/mungedocs/toc.go is used to autogen those?

mikedanese replied:
Not yet unfortunately. Replace it with:
```
**Table of Contents**
<!-- BEGIN MUNGE: GENERATED_TOC -->
<!-- END MUNGE: GENERATED_TOC -->
```

and run ./hack/run-gendocs to autogen

bgrant0607 commented on docs/user-guide/walkthrough/README.md:
```diff
@@ -12,77 +12,176 @@ certainly want the docs that go with that version.</h1>
+- [What's Next?](#whats-next)
+
+
+## Kubectl CLI
```
Would be good to fold this and content from https://github.com/GoogleCloudPlatform/kubernetes/blob/master/docs/getting-started-guides/gce.md#installing-the-kubernetes-command-line-tools-on-your-workstation into https://github.com/GoogleCloudPlatform/kubernetes/blob/master/docs/user-guide/prereqs.md.

But that could be a separate PR.

cc @caesarxuchao

karlkfi replied:
I'm down with having an external kubectl setup user-guide page, but I'm kind of tired of trying to hit a moving target on this PR.

bgrant0607 replied:
Separate PR is fine, as I said.

bgrant0607 commented on docs/user-guide/walkthrough/README.md:
````diff
@@ -12,77 +12,176 @@ certainly want the docs that go with that version.</h1>
+Create a pod containing an nginx server ([pod-nginx.yaml](pod-nginx.yaml)):
+
+```sh
+kubectl create -f examples/walkthrough/pod-nginx.yaml
````
Files have moved

karlkfi replied:
moved since when? I moved them...

bgrant0607 replied:
The path is docs/user-guide/walkthrough/pod-nginx.yaml not examples/walkthrough/pod-nginx.yaml.

@mikedanese may be fixing this in head.

bgrant0607 replied:
#11318 was merged, but this text is new, so it will also have to be updated. cc @krousey

bgrant0607 commented on examples/cmd-exec/README.md:
```diff
@@ -0,0 +1,60 @@
+<!-- END STRIP_FOR_RELEASE -->
+
+<!-- END MUNGE: UNVERSIONED_WARNING -->
+# Command Execution Example
```
Please put this directory under docs/user-guide, also.

karlkfi replied:
what makes this user-guide material and not just a generic example? I made it external so that other docs could link to it, because I've seen the same pattern used in a dozen places.

bgrant0607 replied:
Those dozen places have all been moved to the user guide.

Application/solution-focused examples will remain in /examples. Feature-centric examples have been moved to the user guide.

bgrant0607 commented on docs/user-guide/walkthrough/k8s201.md:
````diff
@@ -12,29 +12,78 @@ certainly want the docs that go with that version.</h1>
-Having already learned about Pods and how to create them, you may be struck by an urge to create many, many pods.  Please do!  But eventually you will need a system to organize these pods into groups.  The system for achieving this in Kubernetes is Labels.  Labels are key-value pairs that are attached to each object in Kubernetes.  Label selectors can be passed along with a RESTful ```list``` req ...
+In order for the kubectl usage examples to work, make sure you have an examples directory locally, either from [a release](https://github.com/GoogleCloudPlatform/kubernetes/releases) or [the source](https://github.com/GoogleCloudPlatform/kubernetes).
+
+Table of Contents
````
Please convert to standard TOC.

bgrant0607 commented:
I'd like to try to get this merged tomorrow.

Please make the suggested changes, rebase, and run hack/run-gendocs.sh

bgrant0607 commented on docs/user-guide/walkthrough/pod-nginx-with-label.yaml:
```diff
@@ -3,7 +3,7 @@ kind: Pod
   name: nginx
   labels:
-    name: nginx
+    app: nginx
```
examples/pod.yaml is being removed. It has been copied to docs/user-guide.

karlkfi replied:
i already removed it and moved it...

caesarxuchao commented on docs/user-guide/walkthrough/README.md:
````diff
@@ -12,77 +12,176 @@ certainly want the docs that go with that version.</h1>
+
+```sh
+export KUBERNETES_MASTER=http://<ip>:<port>/api
+```
````
I'm not sure if this is enough. .kube/config also contains other info like a authentication token. I think the guide [here](https://github.com/GoogleCloudPlatform/kubernetes/blob/master/docs/user-guide/sharing-clusters.md) is more thorough.

karlkfi replied:
You're probably right, KUBERNETES_MASTER is enough only if ssl is disabled and the server is accepting http requests.

karlkfi replied:
Can we punt on this until a kubectl doc extraction? Manual SSL config is kinda complicated...

caesarxuchao replied:
Can we simply point people to https://github.com/GoogleCloudPlatform/kubernetes/blob/master/docs/user-guide/sharing-clusters.md#manually-generating-kubeconfig?

it's not using env var though.

bgrant0607 replied:
We can fix this in a subsequent PR.

karlkfi commented:
I don't understand why shippable & travis are failing:

```
$ PATH=$GOPATH/bin:./third_party/etcd:$PATH ./hack/verify-gendocs.sh
examples/cmd-exec/README.md
----
unversioned-warning:
contents were modified
FAIL: changes needed but not made due to --verify
!!! Error in ./hack/verify-gendocs.sh:63
  '"${mungedocs}" "--verify=true" "--root-dir=${EXAMPLEROOT}"' exited with status 1
Call stack:
  1: ./hack/verify-gendocs.sh:63 main(...)
Exiting with status 1
```

When I run that locally it passes just fine...

karlkfi commented:
Looks like rebase is non-trivial. Someone removed a bunch of trailing whitespace from files in master and it's creating a ton of merge conflicts that I don't have time to deal with this morning.

bgrant0607 commented:
@karlkfi If you'd like someone to take over the PR and get it merged, let us know. We're in a mad scramble to organize the docs for 1.0.

karlkfi commented:
@bgrant0607 I'll look at it this afternoon, but It's beginning to feel like it'll never be done. I'd kinda like to merge what's done and let others iterate on it with further PRs instead of battling the tide of changes.

bgrant0607 commented:
I understand. At minimum, we can't merge until it passes the new checks, though. Currently examples/cmd-exec/README.md fails verify-gendocs. Someone could patch your changes into their fork and work through the issues.

krousey commented:
@karlkfi Hey sorry this has been frustrating. I'm going to patch this in and deal with the remaining headaches with regards to our recent doc changes. I'll let you know when I have a PR. Once we get it checked in, we should be able to iterate any further changes needed.

Thanks again for your contribution!

karlkfi commented:
Superseded by #11336
